Cleaner external project

scottpurdy opened the pull request:
fixes #631 

- Top level CMakeLists.txt delegates to `external/` for externals and `src/` for NuPIC build.
- `external/` uses `find_library` and `find_program` to find externals. If not found, ExternalProject is used to build the libraries. If ExternalProject is used then other targets can depend on it. If it isn't used, then a dummy target is created with the same name for other targets to depend on.
- Next step is to replace pre-built binaries with more ExternalProject calls

@oxtopus - please review

scottpurdy commented:
@oxtopus - I have tested with prebuilt capnproto and without. Please review.

oxtopus commented:
LGTM, but need to get Travis CI passing.  Current failure might be related to version of CMake, so we need to make it clear that the user might need to upgrade, as well as get it upgraded in the CI environment.

scottpurdy commented:
@oxtopus - green now but see the changes I made since you last looked. I added `-DBUILD_SHARED_LIBS=OFF` and `-DCMAKE_CXX_FLAGS=-fPIC -std=c++11 -m64` to the capnp build, fixed up how externals headers are included while handling both pre-installed externals as well as ExternalProject built versions.

breznak commented:
@scottpurdy looks like this could fix https://github.com/numenta/nupic.core/issues/645 
Can you please add/document a simple way to pass capnp library and call just `../../src/CMakeFiles.txt`; 
otherwise it's impossible to rebuild core w/o internet.

scottpurdy commented:
@breznak - added a note to the README in afa69e1. Note that those extra flags are only necessary if your prebuilt version cannot be found in normal system locations.

breznak commented:
@scottpurdy thanks!

Btw,

```
-- Configuring done
CMake Warning (dev) in src/CMakeLists.txt:
  Policy CMP0022 is not set: INTERFACE_LINK_LIBRARIES defines the link
  interface.  Run "cmake --help-policy CMP0022" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "nupic_core_solo" has an INTERFACE_LINK_LIBRARIES property.  This
  should be preferred as the source of the link interface for this library
  but because CMP0022 is not set CMake is ignoring the property and using
the
  link implementation as the link interface instead.

  INTERFACE_LINK_LIBRARIES:


nupic_core_solo;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libyaml-cpp.a;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libyaml.a;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libapr-1.a;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libaprutil-1.a;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libz.a;/usr/local/lib/libkj.a;/usr/local/lib/libcapnp.a;/usr/local/lib/libcapnpc.a;/usr/lib/x86_64-linux-gnu/
libpython2.7.so

  Link implementation:


/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libyaml-cpp.a;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libyaml.a;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libapr-1.a;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libaprutil-1.a;/home/mmm/devel/nupic/nupic.core/external/linux64/lib/libz.a;/usr/local/lib/libkj.a;/usr/local/lib/libcapnp.a;/usr/local/lib/libcapnpc.a;/usr/lib/x86_64-linux-gnu/
libpython2.7.so

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to:
/home/mmm/devel/nupic/nupic.core/build/scripts
```

breznak commented:
And this fixed #645 :+1:

scottpurdy commented:
Thanks for sharing the CMake warning. I am not too sure how to interpret that.
